File: internal/model/errors_test.go
package model_test

import (
	"testing"

	"github.com/wzhejunqiu/data-nexus/internal/model"
)

func TestAppErrorNilReceiver(t *testing.T) {
	var err *model.AppError
	if err.Error() != "" {
		t.Fatalf("expected empty string, got %q", err.Error())
	}
}

func TestAppErrorConstructors(t *testing.T) {
	tests := []struct {
		name string
		err  *model.AppError
		code string
	}{
		{"invalid request", model.ErrInvalidRequest("bad"), "INVALID_REQUEST"},
		{"connection not found", model.ErrConnectionNotFound("id"), "CONNECTION_NOT_FOUND"},
		{"already open", model.ErrConnectionAlreadyOpen("id"), "CONNECTION_ALREADY_OPEN"},
		{"connection open for update", model.ErrConnectionOpen("id"), "CONNECTION_OPEN"},
		{"saved not found", model.ErrSavedNotFound("id"), "SAVED_NOT_FOUND"},
		{"connection failed", model.ErrConnectionFailed("msg"), "CONNECTION_FAILED"},
		{"database locked", model.ErrDatabaseLocked(), "DATABASE_LOCKED"},
		{"table not found", model.ErrTableNotFound("t"), "TABLE_NOT_FOUND"},
		{"sql error", model.ErrSQL("msg"), "SQL_ERROR"},
		{"result too large", model.ErrResultTooLarge(), "RESULT_TOO_LARGE"},
		{"read only", model.ErrReadOnly(), "READ_ONLY"},
		{"dialog cancelled", model.ErrDialogCancelled(), "DIALOG_CANCELLED"},
		{"export cancelled", model.ErrExportCancelled(), "EXPORT_CANCELLED"},
		{"export no stable key", model.ErrExportNoStableKey("t"), "EXPORT_NO_STABLE_KEY"},
		{"invalid path", model.ErrInvalidPath("msg"), "INVALID_PATH"},
		{"internal", model.ErrInternal("msg"), "INTERNAL_ERROR"},
		{"secrets vault locked", model.ErrSecretsVaultLocked(), "SECRETS_VAULT_LOCKED"},
		{"secrets vault not initialized", model.ErrSecretsVaultNotInitialized(), "SECRETS_VAULT_NOT_INITIALIZED"},
		{"secrets vault wrong password", model.ErrSecretsVaultWrongPassword(), "SECRETS_VAULT_WRONG_PASSWORD"},
		{"connection failed with reason", model.ErrConnectionFailedWithReason("failed", "timeout"), "CONNECTION_FAILED"},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			if tc.err.Code != tc.code {
				t.Fatalf("got code %s want %s", tc.err.Code, tc.code)
			}
			if tc.err.Error() == "" {
				t.Fatal("expected non-empty error string")
			}
		})
	}
}

func TestNewAppErrorWithDetails(t *testing.T) {
	err := model.NewAppError("CUSTOM", "detail", map[string]any{"k": "v"})
	if err.Code != "CUSTOM" || err.Details["k"] != "v" {
		t.Fatalf("unexpected error: %+v", err)
	}
}

File: frontend/src/lib/api/errors.test.ts
import { describe, expect, it } from 'vitest'
import i18n from '@/i18n'
import { isDialogCancelled, mapWailsError, translateAppError } from './errors'

const KNOWN_ERROR_CODES = [
  'INVALID_REQUEST',
  'INVALID_PATH',
  'CONNECTION_NOT_FOUND',
  'CONNECTION_ALREADY_OPEN',
  'CONNECTION_FAILED',
  'DATABASE_LOCKED',
  'TABLE_NOT_FOUND',
  'SQL_ERROR',
  'RESULT_TOO_LARGE',
  'READ_ONLY',
  'DIALOG_CANCELLED',
  'SAVED_NOT_FOUND',
  'INTERNAL_ERROR',
] as const

describe('mapWailsError', () => {
  it.each(KNOWN_ERROR_CODES)('parses object shape for %s', (code) => {
    const err = mapWailsError({ code, message: `${code} detail` })
    expect(err.code).toBe(code)
    expect(err.message).toBe(`${code} detail`)
  })

  it.each(KNOWN_ERROR_CODES)('parses string shape for %s', (code) => {
    const err = mapWailsError(`${code}: ${code} detail`)
    expect(err.code).toBe(code)
    expect(err.message).toBe(`${code} detail`)
  })

  it('detects dialog cancelled', () => {
    const err = mapWailsError({ code: 'DIALOG_CANCELLED', message: 'cancelled' })
    expect(isDialogCancelled(err)).toBe(true)
  })

  it('falls back to INTERNAL_ERROR for unknown codes', () => {
    const err = mapWailsError('UNKNOWN_CODE: something broke')
    expect(err.code).toBe('INTERNAL_ERROR')
    expect(err.message).toBe('UNKNOWN_CODE: something broke')
  })

  it('shows SQL error message instead of generic internal text', () => {
    const t = i18n.getFixedT('zh-CN')
    const msg = translateAppError(t, { code: 'SQL_ERROR', message: 'near DESC: syntax error' })
    expect(msg).toBe('near DESC: syntax error')
  })

  it.each([
    ['CONNECTION_FAILED', 'en', 'Failed to open database'],
    ['DATABASE_LOCKED', 'en', 'Database is locked'],
    ['RESULT_TOO_LARGE', 'en', 'Result set exceeds row limit'],
    ['READ_ONLY', 'en', 'read-only'],
    ['CONNECTION_FAILED', 'zh-CN', '无法打开'],
    ['DATABASE_LOCKED', 'zh-CN', '数据库被锁定'],
  ] as const)('translates %s in %s', (code, lng, expected) => {
    const t = i18n.getFixedT(lng)
    const msg = translateAppError(t, { code, message: 'fallback' })
    expect(msg.toLowerCase()).toContain(expected.toLowerCase())
  })

  it('falls back to message for unknown codes', () => {
    const t = i18n.getFixedT('en')
    const msg = translateAppError(t, { code: 'UNKNOWN', message: 'custom' })
    expect(msg).toBe('custom')
  })
})
